src/shared/bootspec.c

```diff
@@ -1046,6 +1052,189 @@ int boot_config_augment_from_loader(
+                _cleanup_(blkid_free_probep) blkid_probe b = NULL;
+                const char *v;
+                const char *devname;
+
+                dev = blkid_verify(cache, dev);
+                if (!dev)
+                        continue;
+
+                devname = blkid_dev_devname(dev);
+
+                errno = 0;
+                b = blkid_new_probe_from_filename(devname);
+                if (!b)
+                        return log_error_errno(errno ?: SYNTHETIC_ERRNO(ENOMEM), "Failed to open file system \"%s\": %m", devname);
+
+                blkid_probe_enable_superblocks(b, 1);
+                blkid_probe_set_superblocks_flags(b, BLKID_SUBLKS_TYPE);
+                blkid_probe_enable_partitions(b, 1);
+                blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
+
+                errno = 0;
+                r = blkid_do_safeprobe(b);
+
+                /*
+                 * Ignore the error because the goal is only to find (if any) partition with
+                 * the same GUID
+                 */
+                r = blkid_probe_lookup_value(b, "PART_ENTRY_TYPE", &v, NULL);
+                if (r != 0)
+                        continue;
+
+                if (devtptr) {
+                        dev_t devid = blkid_probe_get_devno(b);
+                        if (check_parent_by_devid(devid, *devtptr))
+                                continue;
+                }
+
+                if (streq(v, part_uuid))
+                        count++;
+
+                if (dos_part_uuid && streq(v, dos_part_uuid))
+                        count++;
+        }
+
+        return count;
+#else
+        /* we don't have blkid, return a positive result */
+        return 1;
+#endif
+}
+
+static int check_part_type_duplicates_udev(const char *part_uuid, const char *dos_part_uuid, dev_t *devtptr) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        int r;
+        int count = 0;
+        sd_device *d;
+
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_subsystem(e, "block", true);
+        if (r < 0)
+                return r;
+
+        FOREACH_DEVICE(e, d) {
+                const char *v;
+
+                r = sd_device_get_property_value(d, "ID_PART_ENTRY_TYPE", &v);
+                if (r < 0)
+                        continue;
+
+                if (devtptr) {
+                        dev_t devid;
+
+                        r = sd_device_get_devnum(d, &devid);
+                        if (!r && check_parent_by_devid(devid, *devtptr))
+                                continue;
+                }
+
+                if (streq(v, part_uuid))
+                        count++;
+                if (dos_part_uuid && streq(v, dos_part_uuid))
+                        count++;
```

Comment:
so I am not a fan of comparing these things as strings. We generally prefer if we process parsed stuff everywhere. i.e. compare binary stuff, not string representations. Why? because string representations typically may vary for the same value, think uppercasing, prefixing with zeroes and such.

Hence, please pass around an sd_id128_t here for the gpt part type, and a uint8_t for the mbr part type. And parse the strings whenever we get them from udev via sd_id128_from_string() and safe_atox8()

Reply:
Ok